File: src/snippet/python/snl_snippet.py
import sys
import snl

def main() -> int:
  u = snl.SNLUniverse.create()
  db1 = snl.SNLDB.create(u)
  db2 = snl.SNLDB.create(u)
  lib0 = snl.SNLLibrary.create(db1)
  lib1 = snl.SNLLibrary.create(db2)
  lib2 = snl.SNLLibrary.create(lib0)

  model0 = snl.SNLDesign.create(lib1)
  i0 = snl.SNLScalarTerm.create(model0, 'i0')
  i0Net = snl.SNLScalarNet.create(model0, 'i0')
  i0.setNet(i0Net)
  print('Term:', i0, 'in design:', i0.getDesign(), 'connected to:', i0.getNet())
  i1 = snl.SNLScalarTerm.create(model0, 'i0')
  print('Model:', model0) 

  model1 = snl.SNLDesign.create(lib1)
  d0 = snl.SNLDesign.create(lib2)
  print('Design:', d0)

  d0Input0 = snl.SNLScalarTerm.create(d0, 'Input0')
  print('Term:', d0Input0, 'in design:', d0Input0.getDesign())

  i0 = snl.SNLInstance.create(d0, model0, 'i0')
  i1 = snl.SNLInstance.create(d0, model0)
  print('Instance:', i0, ' in design:', i0.getDesign())
  print('Instance:', i1, ' in design:', i1.getDesign())


if __name__ == '__main__':
  sys.exit(main())

File: src/snippet/app/src/SNLSnippet.cpp
#include "SNLUniverse.h"

#include <iostream>

#include "SNLScalarTerm.h"
#include "SNLBusTerm.h"

using namespace SNL;

int main() {
  SNLUniverse::create();
  auto db = SNLDB::create(SNLUniverse::get());
  auto primLib = SNLLibrary::create(db, SNLLibrary::Type::Primitives, SNLName("primitives"));
  auto prim0 = SNLDesign::create(primLib, SNLDesign::Type::Primitive);
  auto prim1 = SNLDesign::create(primLib, SNLDesign::Type::Primitive);

  auto mylib = SNLLibrary::create(db, SNLName("mylib"));
  auto model1 = SNLDesign::create(mylib, SNLName("Model1"));
  SNLScalarTerm::create(model1, SNLTerm::Direction::Input, SNLName("i0"));
  SNLScalarTerm::create(model1, SNLTerm::Direction::Input, SNLName("i1"));
  SNLScalarTerm::create(model1, SNLTerm::Direction::Output, SNLName("o"));
  std::cout << "Model1 terms:" << std::endl;
  for (auto term: model1->getTerms()) {
    std::cout << "  - " << term->getString() << std::endl;
  }

  auto model2 = SNLDesign::create(mylib, SNLName("Model2"));
  SNLBusTerm::create(model2, SNLTerm::Direction::Input, 4, 0, SNLName("i0"));
  SNLScalarTerm::create(model2, SNLTerm::Direction::Input, SNLName("i1"));
  SNLBusTerm::create(model2, SNLTerm::Direction::Output, 31, 0, SNLName("o"));
  std::cout << "Model2 terms:" << std::endl;
  for (auto term: model2->getTerms()) {
    std::cout << "  - " << term->getString() << std::endl;
  }
  
  return 0;
}
